From 9472ec62d714ac477b28392c8a291067b5511f6c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Wed, 10 Sep 2014 14:41:42 +0100
Subject: [PATCH] clip PolyPolygon::Insert on .svm load

valgrind + bff

(cherry picked from commit fd64d444b730f6cb7216dac8f6e3f94b97d7ab60)

Conflicts:
	tools/source/generic/poly2.cxx

Change-Id: Ib3ed8a6e518c0686f8cbeaf021b9ca3a07005032
Reviewed-on: https://gerrit.libreoffice.org/11382
Reviewed-by: Michael Stahl <mstahl@redhat.com>
Tested-by: Michael Stahl <mstahl@redhat.com>
---
 tools/source/generic/poly2.cxx | 8 ++++----
 vcl/source/gdi/metaact.cxx     | 6 ++++++
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/tools/source/generic/poly2.cxx b/tools/source/generic/poly2.cxx
index aa732f4..2641d5e8 100644
--- a/tools/source/generic/poly2.cxx
+++ b/tools/source/generic/poly2.cxx
@@ -219,7 +219,7 @@ void PolyPolygon::Insert( const Polygon& rPoly, sal_uInt16 nPos )
 void PolyPolygon::Remove( sal_uInt16 nPos )
 {
     DBG_CHKTHIS( PolyPolygon, NULL );
-    DBG_ASSERT( nPos < Count(), "PolyPolygon::Remove(): nPos >= nSize" );
+    assert(nPos < Count() && "PolyPolygon::Remove(): nPos >= nSize");
 
     if ( mpImplPolyPolygon->mnRefCount > 1 )
     {
@@ -239,7 +239,7 @@ void PolyPolygon::Remove( sal_uInt16 nPos )
 void PolyPolygon::Replace( const Polygon& rPoly, sal_uInt16 nPos )
 {
     DBG_CHKTHIS( PolyPolygon, NULL );
-    DBG_ASSERT( nPos < Count(), "PolyPolygon::Replace(): nPos >= nSize" );
+    assert(nPos < Count() && "PolyPolygon::Replace(): nPos >= nSize");
 
     if ( mpImplPolyPolygon->mnRefCount > 1 )
     {
@@ -256,7 +256,7 @@ void PolyPolygon::Replace( const Polygon& rPoly, sal_uInt16 nPos )
 const Polygon& PolyPolygon::GetObject( sal_uInt16 nPos ) const
 {
     DBG_CHKTHIS( PolyPolygon, NULL );
-    DBG_ASSERT( nPos < Count(), "PolyPolygon::GetObject(): nPos >= nSize" );
+    assert(nPos < Count() && "PolyPolygon::GetObject(): nPos >= nSize");
 
     return *(mpImplPolyPolygon->mpPolyAry[nPos]);
 }
@@ -617,7 +617,7 @@ Rectangle PolyPolygon::GetBoundRect() const
 Polygon& PolyPolygon::operator[]( sal_uInt16 nPos )
 {
     DBG_CHKTHIS( PolyPolygon, NULL );
-    DBG_ASSERT( nPos < Count(), "PolyPolygon::[](): nPos >= nSize" );
+    assert(nPos < Count() && "PolyPolygon::[](): nPos >= nSize");
 
     if ( mpImplPolyPolygon->mnRefCount > 1 )
     {
diff --git a/vcl/source/gdi/metaact.cxx b/vcl/source/gdi/metaact.cxx
index 6d7ac00..c1c2676 100644
--- a/vcl/source/gdi/metaact.cxx
+++ b/vcl/source/gdi/metaact.cxx
@@ -1192,6 +1192,12 @@ void MetaPolyPolygonAction::Read( SvStream& rIStm, ImplMetaReadData* )
             rIStm >> nIndex;
             Polygon aPoly;
             aPoly.Read( rIStm );
+            if (nIndex >= maPolyPoly.Count())
+            {
+                SAL_WARN("vcl.gdi", "svm contains polygon index " << nIndex
+                         << " outside possible range " << maPolyPoly.Count());
+                continue;
+            }
             maPolyPoly.Replace( aPoly, nIndex );
         }
     }
-- 
2.7.4

